File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/presentation/ViewModelFactory.java
package com.igorganapolsky.vibratingwatchapp.presentation;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.annotation.NonNull;
import com.igorganapolsky.vibratingwatchapp.domain.repo.Repository;
import com.igorganapolsky.vibratingwatchapp.core.timer.ICountdownManager;
import com.igorganapolsky.vibratingwatchapp.presentation.details.TimerDetailsViewModel;
import com.igorganapolsky.vibratingwatchapp.presentation.main.TimerListViewModel;
import com.igorganapolsky.vibratingwatchapp.presentation.settings.SetTimerViewModel;

public class ViewModelFactory implements ViewModelProvider.Factory {

    private static ViewModelFactory factory;

    private final Repository repository;
    private final ICountdownManager countdownManager;

    private ViewModelFactory(Repository repository, ICountdownManager countdownManager) {
        this.repository = repository;
        this.countdownManager = countdownManager;
    }

    public static void initFactory(Repository repository,
                                   ICountdownManager countdownManager) {
        if (factory == null) {
            factory = new ViewModelFactory(repository, countdownManager);
        }
    }

    public static ViewModelProvider.Factory getInstance() {
        return factory;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        ViewModel viewModel;
        if (modelClass.isAssignableFrom(SetTimerViewModel.class)) {
            viewModel = new SetTimerViewModel(repository);
        } else if (modelClass.isAssignableFrom(TimerListViewModel.class)) {
            viewModel = new TimerListViewModel(repository, countdownManager);
        } else if (modelClass.isAssignableFrom(TimerDetailsViewModel.class)) {
            viewModel = new TimerDetailsViewModel(repository, countdownManager);
        } else {
            throw new IllegalStateException("No associated view model with " + modelClass.getName());
        }
        return (T) viewModel;
    }
}

File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/presentation/ViewModelFactory.kt
package com.igorganapolsky.vibratingwatchapp.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.igorganapolsky.vibratingwatchapp.data.ITimersRepository
import com.igorganapolsky.vibratingwatchapp.presentation.home_screen.NewTimerViewModel
import com.igorganapolsky.vibratingwatchapp.presentation.home_screen.ExistingTimersListViewModel
import com.igorganapolsky.vibratingwatchapp.presentation.existing_timer_info.TimerControlViewModel
import com.igorganapolsky.vibratingwatchapp.domain.usecase.ICountdownController

class ViewModelFactory private constructor(
    private val repository: ITimersRepository,
    private val countdownManager: ICountdownController
) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        val viewModel: ViewModel
        if (modelClass.isAssignableFrom(NewTimerViewModel::class.java)) {
            viewModel = NewTimerViewModel(repository)
        } else if (modelClass.isAssignableFrom(ExistingTimersListViewModel::class.java)) {
            viewModel = ExistingTimersListViewModel(repository, countdownManager)
        } else if (modelClass.isAssignableFrom(TimerControlViewModel::class.java)) {
            viewModel = TimerControlViewModel(repository, countdownManager)
        } else {
            throw IllegalStateException("No associated view model with " + modelClass.name)
        }
        return viewModel as T
    }

    companion object {
        private lateinit var factory: ViewModelFactory

        fun initFactory(repository: ITimersRepository, countdownController: ICountdownController) {
            factory = ViewModelFactory(repository, countdownController)
        }

        val instance: ViewModelProvider.Factory?
            get() = factory
    }
}
